legend-depot-core-tracing/src/main/java/org/finos/legend/depot/tracing/resources/BaseResource.java

```diff
@@ -97,8 +101,48 @@ protected <T> T handle(String resourceAPIMetricName, String label, Supplier<T> s
         }
     }
 
+    protected <T> Response handle(String resourceAPIMetricName, String label, Supplier<T> supplier, Request request, String groupId, String artifactId, String versionId)
+    {
+        EntityTag serverTag = generateETag(groupId, artifactId, versionId);
+        if (request.evaluatePreconditions(serverTag) != null)
+        {
+            return Response.noContent().status(Response.Status.NOT_MODIFIED).build();
+        }
+        try
+        {
+            return Response.ok(handle(resourceAPIMetricName, label, supplier)).tag(serverTag).build();
+        }
+        catch (Exception e)
+        {
+            throw new WebApplicationException(e.getMessage(), e);
+        }
+    }
+
     protected <T> T handle(String label, Supplier<T> supplier)
     {
         return handle(label, label, supplier);
     }
+
+    protected <T> Response handle(String label, Supplier<T> supplier, Request request, String groupId, String artifactId, String versionId)
+    {
+        EntityTag serverTag = generateETag(groupId, artifactId, versionId);
+        if (request.evaluatePreconditions(serverTag) != null)
+        {
+            return Response.noContent().status(Response.Status.NOT_MODIFIED).build();
+        }
+        try
+        {
+            return Response.ok(handle(label, supplier)).tag(serverTag).build();
+        }
+        catch (Exception e)
+        {
+            throw new WebApplicationException(e.getMessage(), e);
+        }
+    }
+
+    protected EntityTag generateETag(String groupId, String artifactId, String versionId)
```

Comment:
this method has to be mroe genercis ,,as in more params determine uniqueness of response